vizro-core/tests/unit/vizro/models/_action/test_action.py

```diff
@@ -0,0 +1,205 @@
+        with pytest.raises(Exception):
+            Action(function=test_action_function, inputs=inputs, outputs=[])
+
+    @pytest.mark.parametrize(
+        "outputs",
+        [
+            None,
+            "string",
+            "component_property",
+            "compo-nent.property",
+            "component.property_1",
+        ],
+    )
+    def test_outputs_invalid(self, outputs, test_action_function):
+        with pytest.raises(Exception):
+            Action(function=test_action_function, inputs=[], outputs=outputs)
+
+
+class TestActionBuild:
+    """Tests action build method."""
+
+    def test_build_no_returned_components_for_custom_action(self, test_action_function):
+        action = Action(function=test_action_function)
+        result = action.build()
+        expected = []
+        assert result == expected
+
+
+class TestActionPrivateMethods:
+    """Test action private methods."""
+
+    def test_get_callback_mapping_no_inputs_no_outputs(self, test_action_function):
+        action = Action(id="action_test", function=test_action_function)
+        callback_inputs, callback_outputs, action_components = action._get_callback_mapping()
+        assert callback_inputs == {
+            "trigger": dash.Input({"action_name": "action_test", "type": "action_trigger"}, "data")
+        }
+        assert callback_outputs == {"action_finished": dash.Output("action_finished", "data")}
+        assert action_components == []
+
+    @pytest.mark.parametrize(
+        "inputs, outputs",
+        [
+            (["component_1.property"], ["component_1.property"]),
+            (["component_1.property", "component_2.property"], ["component_1.property", "component_2.property"]),
+        ],
+    )
+    def test_get_callback_mapping_with_inputs_and_outputs(self, inputs, outputs, test_action_function):
+        action = Action(
+            function=test_action_function,
+            inputs=inputs,
+            outputs=outputs,
+        )
+        callback_inputs, callback_outputs, action_components = action._get_callback_mapping()
+        assert callback_inputs == {
+            **{
+                f'{input.split(".")[0]}_{input.split(".")[1]}': dash.State(input.split(".")[0], input.split(".")[1])
+                for input in inputs
+            },
+            "trigger": dash.Input({"action_name": action.id, "type": "action_trigger"}, "data"),
+        }
+        assert callback_outputs == {
+            **{
+                f'{output.split(".")[0]}_{output.split(".")[1]}': dash.Output(
+                    output.split(".")[0], output.split(".")[1], allow_duplicate=True
+                )
+                for output in outputs
+            },
+            "action_finished": dash.Output("action_finished", "data"),
+        }
+        assert action_components == []
+
+    @pytest.mark.parametrize(
+        "custom_action_function, callback_context_outputs_grouping, expected_function_return_value",
```

Comment:
I found this unintuitive to read. Source of my confusion (and maybe we can make this better if you agree that the points are valid):

1. `custom_action_function` does not look like a function until realizing later how it handle's `requests`. Maybe a renaming might be useful, smth like `custom_action_function_mock_return`
2. Similar for the `callback_context_outputs_grouping`, solution potentially similar to above - for filter tests we chose e.g. `callback_context_filter_continent` so that the parametrization is a lot more intuitive --> can we achieve a name that suggests what this context is parametrized for?